src/librustc_mir/transform/promote_consts.rs

```diff
@@ -207,14 +207,26 @@ impl<'tcx> Visitor<'tcx> for Collector<'_, 'tcx> {
             Rvalue::Ref(..) => {
                 self.candidates.push(Candidate::Ref(location));
             }
-            Rvalue::Repeat(elem, _)
+            Rvalue::Repeat(elem, n)
                 if self.ccx.tcx.features().const_in_array_repeat_expressions =>
             {
-                if !elem.ty(&self.ccx.body.local_decls, self.ccx.tcx).is_copy_modulo_regions(
+                let is_copy = elem.ty(&self.ccx.body.local_decls, self.ccx.tcx).is_copy_modulo_regions(
                     self.ccx.tcx,
                     self.ccx.param_env,
                     self.span,
-                ) {
+                );
+                let n = n.try_eval_usize(self.ccx.tcx, self.ccx.param_env);
+                let length_requires_copy_or_promotion = match n {
+                    // Unevaluable (e.g. too generic) -> assume > 1.
+                    None => true,
+                    // A single element does not need copying, we can just use the given element.
+                    // Zero elements don't even need that one element.
+                    Some(0..=1) => false,
```

Comment:
Ah, this makes sense. :D

Reply:
It also seems quite fragile (different compiler passes independently making the same decision) but I am not sure if there is a way around that.